### PR TITLE
[Broker] Fix call sync method in onPoliciesUpdate method

### PR DESCRIPTION
**Pull request by nodece**
Signed-off-by: Zixuan Liu <nodeces@gmail.com>

### Motivation

Remove call sync method in `PersistentTopic#onPoliciesUpdate()` and `NonPersistentTopic#onPoliciesUpdate()` methods, avoid block the thread.

### Modifications

- Use async method instead of sync method

### Documentation

- [x] `no-need-doc`

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/nonpersistent/NonPersistentTopic.java [hunk 1]
```diff
@@ -1003,13 +1004,19 @@ public void checkDeduplicationSnapshot() {
         isAllowAutoUpdateSchema = data.is_allow_auto_update_schema;
         schemaValidationEnforced = data.schema_validation_enforced;
 
-        producers.values().forEach(producer -> {
-            producer.checkPermissions();
-            producer.checkEncryption();
-        });
-        subscriptions.forEach((subName, sub) -> sub.getConsumers().forEach(Consumer::checkPermissions));
+        List<CompletableFuture<Void>> producerCheckFutures = new ArrayList<>();
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/nonpersistent/NonPersistentTopic.java:1007]:
```suggestion
        List<CompletableFuture<Void>> producerCheckFutures = new ArrayList<>(producers.size());
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 2]
```diff
@@ -2436,39 +2436,49 @@ private boolean shouldTopicBeRetained() {
...
-        // update rate-limiter if policies updated
-        if (this.dispatchRateLimiter.isPresent()) {
-            if (!topicPolicies.isPresent() || !topicPolicies.get().isDispatchRateSet()) {
-                dispatchRateLimiter.get().onPoliciesUpdate(data);
-            }
-        }
-        if (this.subscribeRateLimiter.isPresent()) {
-            subscribeRateLimiter.get().onPoliciesUpdate(data);
-        }
+        List<CompletableFuture<Void>> producerCheckFutures = new ArrayList<>();
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2439]:
```suggestion
        List<CompletableFuture<Void>> producerCheckFutures = new ArrayList<>(producers.size());
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 3]
```diff
@@ -2436,39 +2436,49 @@ private boolean shouldTopicBeRetained() {
...
+        producers.values().forEach(producer -> producerCheckFutures.add(
+                producer.checkPermissionsAsync().thenRun(producer::checkEncryption)));
+
+        return FutureUtil.waitForAll(producerCheckFutures).thenCompose((__)->{
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2443]:
```suggestion
        return FutureUtil.waitForAll(producerCheckFutures).thenCompose( (__) -> {
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 4]
```diff
@@ -2436,39 +2436,49 @@ private boolean shouldTopicBeRetained() {
...
+            List<CompletableFuture<Void>> subscriptionCheckFutures = new ArrayList<>();
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2444]:
```suggestion
            List<CompletableFuture<Void>> subscriptionCheckFutures = new ArrayList<>(subscriptions.size());
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 5]
```diff
@@ -2436,39 +2436,49 @@ private boolean shouldTopicBeRetained() {
...
+            subscriptions.forEach((subName, sub) -> {
+                List<CompletableFuture<Void>> consumerCheckFutures = new ArrayList<>();
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2446]:
```suggestion
                List<CompletableFuture<Void>> consumerCheckFutures = new ArrayList<>(sub.getConsumers().size());
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 6]
```diff
@@ -2436,39 +2436,49 @@ private boolean shouldTopicBeRetained() {
...
+                    // the topic-level policy.
+                    if (dispatcher != null && (!topicPolicies.isPresent() || !topicPolicies.get()
+                            .isSubscriptionDispatchRateSet())) {
+                        dispatcher.getRateLimiter()
+                                .ifPresent(rateLimiter -> rateLimiter.onPoliciesUpdate(data));
+                    }
+                }));
+            });
+
+            return FutureUtil.waitForAll(subscriptionCheckFutures).thenCompose((___)->{
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2460]:
```suggestion
            return FutureUtil.waitForAll(subscriptionCheckFutures).thenCompose((___) -> {
```

**Review by codelipenghui — COMMENTED** [review]
Great job!

**Review by codelipenghui — APPROVED** [review]

**Review by Shoothzj — APPROVED** [review]

**Review by 315157973 — APPROVED** [review]

**Review by RobertIndie — APPROVED** [review]

**Review by congbobo184 — APPROVED** [review]

**Comment by xiaobing0508** [comment]
Which version fixes this bug?

**Comment by xiaobing0508** [comment]
Is there a binary package?

**Pull request merged**